browser/net/decentralized_dns_network_delegate_helper.cc

```diff
@@ -0,0 +1,106 @@
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/net/decentralized_dns_network_delegate_helper.h"
+
+#include <vector>
+
+#include "net/base/net_errors.h"
+
+#include "brave/browser/brave_wallet/brave_wallet_service_factory.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_service.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_utils.h"
+#include "brave/components/brave_wallet/browser/eth_json_rpc_controller.h"
+#include "brave/components/decentralized_dns/constants.h"
+#include "brave/components/decentralized_dns/utils.h"
+#include "chrome/browser/browser_process.h"
+#include "content/public/browser/browser_context.h"
+
+namespace decentralized_dns {
+
+namespace {
+
+std::string GetValue(const std::vector<std::string>& arr, RecordKeys key) {
+  return arr[static_cast<size_t>(key)];
+}
+
+}  // namespace
+
+int OnBeforeURLRequest_DecentralizedDnsPreRedirectWork(
+    const brave::ResponseCallback& next_callback,
+    std::shared_ptr<brave::BraveRequestInfo> ctx) {
+  if (!ctx->browser_context || !IsDecentralizedDnsEnabled() ||
+      ctx->browser_context->IsOffTheRecord() || !g_browser_process) {
+    return net::OK;
+  }
+
+  if (IsUnstoppableDomainsTLD(ctx->request_url) &&
+      IsUnstoppableDomainsResolveMethodEthereum(
+          g_browser_process->local_state())) {
+    auto* service = BraveWalletServiceFactory::GetInstance()->GetForContext(
+        ctx->browser_context);
+    if (!service) {
+      return net::OK;
+    }
+
+    service->controller()->UnstoppableDomainsProxyReaderGetMany(
+        kProxyReaderContractAddress, ctx->request_url.host(),
+        std::vector<std::string>(std::begin(kRecordKeys),
+                                 std::end(kRecordKeys)),
+        base::BindOnce(&OnBeforeURLRequest_DecentralizedDnsRedirectWork,
+                       next_callback, ctx));
+
+    return net::ERR_IO_PENDING;
+  }
+
+  return net::OK;
+}
+
+void OnBeforeURLRequest_DecentralizedDnsRedirectWork(
+    const brave::ResponseCallback& next_callback,
+    std::shared_ptr<brave::BraveRequestInfo> ctx,
+    bool success,
+    const std::string& result) {
+  if (!success) {
+    if (!next_callback.is_null())
+      next_callback.Run();
+    return;
+  }
+
+  std::vector<std::string> output;
+  size_t offset = 2 /* len of "0x" */ + 64 /* len of offset to array */;
+  if (offset > result.size() ||
+      !brave_wallet::DecodeStringArray(result.substr(offset), &output)) {
+    if (!next_callback.is_null())
+      next_callback.Run();
+    return;
+  }
+
+  // Redirect to ipfs URI if content hash is set, otherwise, fallback to the
+  // set redirect URL. If no records available to use, do nothing.
+  //
+  // TODO(jocelyn): Do not fallback to the set redirect URL if dns.A or
+  // dns.AAAA is not empty once we support the classical DNS records case.
+  std::string ipfs_uri = GetValue(output, RecordKeys::DWEB_IPFS_HASH);
```

Comment:
would be extremely helpful to have a link to the protocol description

Reply:
1cdb4f081decc67ac46ac53ed9bee83fcf3ebb34